src/middlewared/middlewared/alert/source/m_series_nvdimm_firmware_freebsd.py

```diff
@@ -0,0 +1,38 @@
+# -*- coding=utf-8 -*-
+import logging
+
+from middlewared.alert.base import Alert, AlertClass, AlertCategory, AlertLevel, AlertSource
+from middlewared.service_exception import CallError
+
+logger = logging.getLogger(__name__)
+
+
+class TrueNASMNVDIMMFirmwareVersionAlertClass(AlertClass):
+    category = AlertCategory.HARDWARE
+    level = AlertLevel.EMERGENCY
+    title = "Invalid NVDIMM Firmware Version"
+    text = (
+        "Your NVDIMM device (nvdimm%(index)d) is using an invalid firmware version %(version)s which can cause data "
+        "loss. Please contact support."
+    )
+
+    products = ("ENTERPRISE",)
+    hardware = True
+
+
+class TrueNASMNVDIMMFirmwareVersionAlertSource(AlertSource):
+    products = ("ENTERPRISE",)
+
+    async def check(self):
+        if (await self.middleware.call("truenas.get_chassis_hardware")).startswith("TRUENAS-M"):
+            for nvdimm in await self.middleware.call("enterprise.m_series_nvdimm"):
+                size_to_version = {16: "2.2", 32: "2.4"}
+                if nvdimm["size"] not in size_to_version:
+                    raise CallError(f"Unknown NVDIMM size: {nvdimm['size']}")
```

Comment:
Does it make any sense to raise a CallError here? Wouldnt just a log be better?

Reply:
It will create an alert and user will know immediately that his hardware is weird. Log will just get lost.